### PR TITLE
Allow pom download failures

### LINKED ISSUE #4687: openrewrite maven recipes shoudn't try to resolve every single pom
Hello,

I have projects with external dependencies that have been installed manually into a local maven repository.
These artefacts don't have a pom in the repository.
On these projects 'mvn clean install' succeeds and there is only a warning about the missing pom :
```log
[WARNING] The POM for xxxxxxx is missing, no dependency information available
```

However when I try to run any openrewrite maven recipes on this kind of project, it fails with :
```log
[ERROR] </project>org.openrewrite.maven.MavenDownloadingException: Unable to download POM:xxxxx Tried repositories:
[ERROR] https://repo.maven.apache.org/maven2: HTTP 404
[ERROR]   org.openrewrite.maven.internal.MavenPomDownloader.download(MavenPomDownloader.java:636)
[ERROR]   org.openrewrite.maven.tree.ResolvedPom.resolveDependencies(ResolvedPom.java:929)
[ERROR]   org.openrewrite.maven.tree.ResolvedPom.resolveDependencies(ResolvedPom.java:851)
[ERROR]   org.openrewrite.maven.tree.MavenResolutionResult.resolveDependencies(MavenResolutionResult.java:175)
```

Why openrewrite maven recipes would fail when a maven build succeeds with warnings ?
How to fix this ?

You can reproduce this bug with this minimal project : [project.zip](https://github.com/user-attachments/files/17821254/project.zip)

 Just install any jar locally with : 
```shell
mvn install:install-file -Dfile=<any jar file> -DgroupId=bug.report -DartifactId=somedep -Dversion=1.0 -Dpackaging=jar -DgeneratePom=false
```

Then 'mvn clean install' will give : 
```log
[INFO] Scanning for projects...
[INFO] 
[INFO] -------------------------< bug.report:somejar >-------------------------
[INFO] Building somejar 1.0-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[WARNING] The POM for bug.report:somedep:jar:1.0 is missing, no dependency information available
[INFO] 
[INFO] --- maven-clean-plugin:2.5:clean (default-clean) @ somejar ---
[INFO] Deleting /home/stef/repos_github/bugreport/target
[INFO] 
[INFO] --- maven-resources-plugin:2.6:resources (default-resources) @ somejar ---
[WARNING] Using platform encoding (UTF-8 actually) to copy filtered resources, i.e. build is platform dependent!
[INFO] skip non existing resourceDirectory /home/stef/repos_github/bugreport/src/main/resources
[INFO] 
[INFO] --- maven-compiler-plugin:3.8.1:compile (default-compile) @ somejar ---
[INFO] Changes detected - recompiling the module!
[WARNING] File encoding has not been set, using platform encoding UTF-8, i.e. build is platform dependent!
[INFO] Compiling 1 source file to /home/stef/repos_github/bugreport/target/classes
[INFO] 
[INFO] --- maven-resources-plugin:2.6:testResources (default-testResources) @ somejar ---
[WARNING] Using platform encoding (UTF-8 actually) to copy filtered resources, i.e. build is platform dependent!
[INFO] skip non existing resourceDirectory /home/stef/repos_github/bugreport/src/test/resources
[INFO] 
[INFO] --- maven-compiler-plugin:3.8.1:testCompile (default-testCompile) @ somejar ---
[INFO] No sources to compile
[INFO] 
[INFO] --- maven-surefire-plugin:2.12.4:test (default-test) @ somejar ---
[INFO] No tests to run.
[INFO] 
[INFO] --- maven-jar-plugin:2.4:jar (default-jar) @ somejar ---
[INFO] Building jar: /home/stef/repos_github/bugreport/target/somejar-1.0-SNAPSHOT.jar
[INFO] 
[INFO] --- maven-install-plugin:2.4:install (default-install) @ somejar ---
[INFO] Installing /home/stef/repos_github/bugreport/target/somejar-1.0-SNAPSHOT.jar to /home/stef/.m2/repository/bug/report/somejar/1.0-SNAPSHOT/somejar-1.0-SNAPSHOT.jar
[INFO] Installing /home/stef/repos_github/bugreport/pom.xml to /home/stef/.m2/repository/bug/report/somejar/1.0-SNAPSHOT/somejar-1.0-SNAPSHOT.pom
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  1.105 s
[INFO] Finished at: 2024-11-19T21:36:09+01:00
```

Then running 'mvn org.openrewrite.maven:rewrite-maven-plugin:run -Drewrite.activeRecipes=org.openrewrite.maven.cleanup.PrefixlessExpressions`will fail with : 
```log
[ERROR] Failed to execute goal org.openrewrite.maven:rewrite-maven-plugin:5.45.0:run (default-cli) on project somejar: Failed to parse or resolve the Maven POM file or one of its dependencies; We can not reliably continue without this information.: Failed to parse or resolve the Maven POM file or one of its dependencies; We can not reliably continue without this information. 
[ERROR] <!--~~(Unable to download POM: bug.report:somedep:1.0. Tried repositories:
[ERROR] https://repo.maven.apache.org/maven2: HTTP 404)~~>--><project xmlns="http://maven.apache.org/POM/4.0.0"
[ERROR]          xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
[ERROR]          xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
[ERROR]     <modelVersion>4.0.0</modelVersion>
[ERROR] 
[ERROR]     <groupId>bug.report</groupId>
[ERROR]     <artifactId>somejar</artifactId>
[ERROR]     <version>1.0-SNAPSHOT</version>
[ERROR] 
[ERROR]     <properties>
[ERROR]         <maven.compiler.source>1.8</maven.compiler.source>
[ERROR]         <maven.compiler.target>1.8</maven.compiler.target>
[ERROR]     </properties>
[ERROR] 
[ERROR]     <dependencies>
[ERROR]         <!-- any jar installed with :
[ERROR]         mvn install:install-file -Dfile=<any jar file> -DgroupId=bug.report -DartifactId=somedep -Dversion=1.0 -Dpackaging=jar -DgeneratePom=false
[ERROR]         -->
[ERROR]         <dependency>
[ERROR]             <groupId>bug.report</groupId>
[ERROR]             <artifactId>somedep</artifactId>
[ERROR]             <version>1.0</version>
[ERROR]         </dependency>
[ERROR]     </dependencies>
[ERROR] 
[ERROR]     <build>
[ERROR]         <plugins>
[ERROR]             <plugin>
[ERROR]                 <groupId>org.apache.maven.plugins</groupId>
[ERROR]                 <artifactId>maven-compiler-plugin</artifactId>
[ERROR]                 <version>3.8.1</version>
[ERROR]             </plugin>
[ERROR]         </plugins>
[ERROR]     </build>
[ERROR] </project>org.openrewrite.maven.MavenDownloadingException: Unable to download POM: bug.report:somedep:1.0. Tried repositories:
[ERROR] https://repo.maven.apache.org/maven2: HTTP 404
[ERROR]   org.openrewrite.maven.internal.MavenPomDownloader.download(MavenPomDownloader.java:636)
[ERROR]   org.openrewrite.maven.tree.ResolvedPom.resolveDependencies(ResolvedPom.java:929)
[ERROR]   org.openrewrite.maven.tree.ResolvedPom.resolveDependencies(ResolvedPom.java:851)
[ERROR]   org.openrewrite.maven.tree.MavenResolutionResult.resolveDependencies(MavenResolutionResult.java:175)
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
```

### PR DESCRIPTION
**Pull request by adastraperangusta**
## What's changed?
This PR adds a `org.openrewrite.allowPomDownloadFailure` property , inspired by the existing `org.openrewrite.test.readMavenSettingsFromDisk` property. When set to true, the MavenPomDowloader class doesn't throw an exception when it doesn't find a suitable pom to download for a dependency, and instead returns a minimal object.

## What's your motivation?
Some maven projects are correct (mave wise) but don't have all the pom's for their dependencies (see https://github.com/openrewrite/rewrite/issues/4687), which prevents maven recipes to work on these kind of project.
- Fix #4687

## Have you considered any alternatives or workarounds?
This is a workaround, I have the feeling that the real solution would be to use maven libraries to build the maven model instead of re-implementing it. But it would be quite an heavy change.

### Checklist
- [X] I've added unit tests to cover both positive and negative cases
- [X] I've read and applied the [recipe conventions and best practices](https://docs.openrewrite.org/authoring-recipes/recipe-conventions-and-best-practices)
- [X] I've used the IntelliJ IDEA auto-formatter on affected files

### CODE REVIEW

**Review by timtebeek — COMMENTED** [review]
Thanks for exploring the fix here @adastraperangusta !

Saving others a click by copying the test failure here:
```
MavenPomDownloaderTest > WithNativeHttpURLConnectionAndTLS > mirrorsOverrideRepositoriesInPom() FAILED
    org.opentest4j.AssertionFailedError: 
    expected: "https://artifactory.moderne.ninja/artifactory/moderne-cache/"
     but was: "https://artifactory.moderne.ninja/artifactory/moderne-cache"
        at app//org.openrewrite.maven.internal.MavenPomDownloaderTest$WithNativeHttpURLConnectionAndTLS.mirrorsOverrideRepositoriesInPom(MavenPomDownloaderTest.java:251)
```

File: rewrite-maven/src/test/java/org/openrewrite/maven/internal/MavenPomDownloaderTest.java [hunk 1]
```diff
@@ -573,6 +574,42 @@ void skipsLocalInvalidArtifactsEmptyJar(@TempDir Path localRepository) throws IO
...
+                 <project>
+                   <groupId>com.some</groupId>
+                   <artifactId>some-artifact</artifactId>
+                   <version>1</version>
+                 </project>
+                """
+            );
+            Path localJar = localRepository.resolve("com/some/some-artifact/1/some-artifact-1.jar");
+            Files.writeString(localJar, "");
+
```
Comment (timtebeek) [inline rewrite-maven/src/test/java/org/openrewrite/maven/internal/MavenPomDownloaderTest.java:597]: It looks like none of these lines are necessary to make the test pass; Is there any reason to keep these then? Should the logic reflect perhaps that we're ok to continue if only the `.jar` is present?
```suggestion
```

**Comment by adastraperangusta** [comment]
Hello,

I have applied the changes and rebased the PR.
Everything seems OK now.

**Review by timtebeek — COMMENTED** [review]
Thanks for the work here so far @adastraperangusta ; I worry that we might be a little too permissive with this change, as also pointed out on [this comment](https://github.com/openrewrite/rewrite/pull/4738#discussion_r1869210909). Right now we allow any non existing GAV, whereas it seems Maven perhaps only allows a missing pom.xml if the matching .jar is present. What are your thoughts on replicating that here?

**Comment by adastraperangusta** [comment]
Yes, I understand that. It's hack, which doesn't bother that much because when I try to apply a recipe based on the MavenParser, the project is OK maven wise. And no "random" dependency would go missing.

However, I'm ready to give a try to check if jar is present before allowing download failure or not, but how is it possible to detect this when we are in org.openrewrite.maven.internal.MavenPomDownloader#download() method ?
It's beyond my current understanding of the code.

**Comment by timtebeek** [comment]
Appreciate your willingness to help out! I don't quite know just yet how we'd handle that; perhaps we can throw and catch a specific exception when the pom can not be found, and catch that some place where it's easier to check the .jar before rethrowing?

**Comment by adastraperangusta** [comment]
I have added test cases for expected behaviour :
- trying to download a pom without a corresponding jar should fail
- trying to download a pom with a corresponding jar shouldn't fail

The first test fails since the logic isn't implemented yet.

Now after reading the code I have the feeling that the only mean to achieve this is to double each pom request. Everytime a request to the pom URI fails for a repo, we should try the same request for the jar's URI.
If jar is found, then don't throw failure.
BTW by doing like this we can drop the configuration property.

**Comment by timtebeek** [comment]
Thanks for the tests already! Also let me know if you'd like me to step in; the tests will help me explore alternatives more easily.

**Comment by timtebeek** [comment]
A quick note to let you know here that I'm exploring options here to handle this without any option to enable, but merely catching the failure to download the pom, and then looking to see if the `.jar` can be accessed, and if so (and only then) continue as normal.

**Comment by adastraperangusta** [comment]
Hello,

I have update the logic for the 'file' URI type so that when pom download is failed but a local jar is found, the RawPom is built from the jar information which allow to return the needed pom information.
With this logic the property isn't needed anymore, so I removed it.

The 'http' scheme remains to be improved the same way. I have added the corresponding tests cases.

Regards,

**Comment by timtebeek** [comment]
Thanks @adastraperangusta ! I'll try to rework what you've done into the approach I had in mind, as we also want to account for jars that can be downloaded; especially appreciate the tests you have written; they will make it easier to refactor. 🙏🏻

**Comment by adastraperangusta** [comment]
I have added the logic for remote pom branch.
The download() method is growing complicated and surely needs some refactor.

Regards,

**Comment by timtebeek** [comment]
Thanks again @adastraperangusta ; I'll take it from here to avoid conflicts. I've started a parallel effort in this PR to compare solutions:
- https://github.com/openrewrite/rewrite/pull/4777

**Review by timtebeek — APPROVED** [review]
Thanks a lot for the thoughtful work here @adastraperangusta ! Especially appreciate you writing tests for each combination of with/without jar and using file or http transfer.

I've added some light polish like switching to a separate method that skips downloading the jar bytes, as we're only interested in whether the response was successful. I've also added some tests in ResolvedPomTest to ensure we only attempt to reach out once there.

**Pull request merged**